java/lance-jni/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod blocking_dataset;
+mod jni_helpers;
+
+use crate::blocking_dataset::BlockingDataset;
+use arrow::ffi_stream::{ArrowArrayStreamReader, FFI_ArrowArrayStream};
+use jni::objects::{JMap, JObject, JString};
+use jni::sys::{jint, jlong};
+use jni::JNIEnv;
+use lance::dataset::{WriteMode, WriteParams};
+
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_Dataset_writeWithFFIStream<'local>(
+    mut env: JNIEnv<'local>,
+    _obj: JObject,
+    arrow_array_stream_addr: jlong,
+    path: JString,
+    params: JObject,
+) -> JObject<'local> {
+    let path_str = match jni_helpers::extract_path_str(&mut env, &path) {
+        Ok(value) => value,
+        Err(err_str) => {
+            env.throw_new("java/lang/IllegalArgumentException", err_str)
+                .expect("Error throwing exception");
+            return JObject::null();
+        }
+    };
+
+    let write_params = match extract_write_params(&mut env, &params) {
+        Ok(value) => value,
+        Err(err_str) => {
+            env.throw_new("java/lang/IllegalArgumentException", err_str)
+                .expect("Error throwing exception");
+            return JObject::null();
+        }
+    };
+
+    unsafe {
+        match ArrowArrayStreamReader::from_raw(arrow_array_stream_addr as *mut FFI_ArrowArrayStream)
+        {
```

Comment:
Let's make sure when we use `unsafe` blocks, we only put the minimal unsafe parts there and leave a comment why the calls are safe.

```rust
let stream_ptr = arrow_array_stream_addr as *mut FFI_ArrowArrayStream;
// SAFETY: the pointer is recieved directly from Java's 
// ArrowArrayStream.memory_address(), which guarantees to return a non-null
// valid pointer.
let stream = unsafe { ArrowArrayStreamReader::from_raw(stream_ptr) };
match stream { .... }
```

Reply:
+1

Reply:
@eddyxu This is still outstanding

Reply:
Fixed